contracts/randdrop/src/contract.rs

```diff
@@ -0,0 +1,843 @@
+use cosmwasm_std::{
+    ensure_eq, entry_point, to_binary, Addr, Attribute, BankMsg, Coin, CosmosMsg, Deps, DepsMut,
+    Env, HexBinary, MessageInfo, Order, QueryResponse, Response, StdResult, Uint128, WasmMsg,
+};
+use nois::{NoisCallback, ProxyExecuteMsg};
+use sha2::{Digest, Sha256};
+
+use crate::error::ContractError;
+use crate::msg::{
+    ConfigResponse, ExecuteMsg, InstantiateMsg, IsClaimedResponse, IsWinnerResponse,
+    MerkleRootResponse, QueryMsg, ResultsResponse,
+};
+use crate::state::{
+    Config, NoisProxy, ParticipantData, CLAIMED, CONFIG, MERKLE_ROOT, PARTICIPANTS,
+};
+
+/// The winning chance is 1/AIRDROP_ODDS
+const AIRDROP_ODDS: u64 = 3;
+
+#[entry_point]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    _info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    let manager = deps.api.addr_validate(&msg.manager)?;
+    let randdrop_denom = msg.randdrop_denom;
+    let nois_proxy_address = deps
+        .api
+        .addr_validate(&msg.nois_proxy_address)
+        .map_err(|_| ContractError::InvalidProxyAddress)?;
```

Comment:
Why do we have a special error case for invalid proxy but not invalid manager? Not sure what best practice is here. I think both ways are fine.

Reply:
Gonna add a new error for Invalid manager Address